tests/test_replication.py

```diff
@@ -0,0 +1,173 @@
+import json
+
+from mock import Mock
+from sydent.threepid import ThreepidAssociation
+from sydent.threepid.signer import Signer
+from tests.utils import make_request, make_sydent
+from twisted.web.client import Response
+from twisted.internet import defer
+from twisted.trial import unittest
+
+
+class ReplicationTestCase(unittest.TestCase):
+    """Test that a Sydent can correctly replicate data with another Sydent"""
+
+    def setUp(self):
+        # Create a new sydent
+        config = {
+            "crypto": {
+                "ed25519.signingkey": "ed25519 0 FJi1Rnpj3/otydngacrwddFvwz/dTDsBv62uZDN2fZM"
+            }
+        }
+        self.sydent = make_sydent(test_config=config)
+
+        # Create a fake peer to replicate to.
+        peer_public_key_base64 = "+vB8mTaooD/MA8YYZM8t9+vnGhP1937q2icrqPV9JTs"
+
+        # Inject our fake peer into the database.
+        cur = self.sydent.db.cursor()
+        cur.execute(
+            "insert into peers (name, port, lastSentVersion, active) VALUES (?, ?, ?, ?)",
+            ("fake.server", 1234, 0, 1)
+        )
+        cur.execute(
+            "insert into peer_pubkeys (peername, alg, key) VALUES (?, ?, ?)",
+            ("fake.server", "ed25519", peer_public_key_base64)
+        )
+
+        self.sydent.db.commit()
+
+        # Build some fake associations.
+        self.assocs = []
+        assoc_count = 150
+        for i in range(assoc_count):
+            assoc = ThreepidAssociation(
+                medium="email",
+                address="bob%d@example.com" % i,
+                lookup_hash=None,
+                mxid="@bob%d:example.com" % i,
+                ts=(i * 10000),
+                not_before=0,
+                not_after=99999999999,
+            )
+            self.assocs.append(assoc)
+
+    def test_incoming_replication(self):
+        """Impersonate a peer that sends a replication push to Sydent, then checks that it
+        accepts the payload and saves it correctly.
+        """
+        self.sydent.run()
+
+        # Configure the Sydent to impersonate. We need to use "fake.server" as the
+        # server's name because that's the name the recipient Sydent has for it, and
+        # because the certificate's common name is for this name.
+        config = {
+            "general": {
+                "server.name": "fake.server"
+            },
+            "crypto": {
+                "ed25519.signingkey": "ed25519 0 b29eXMMAYCFvFEtq9mLI42aivMtcg4Hl0wK89a+Vb6c"
+            }
+        }
+
+        fake_sydent = make_sydent(config)
```

Comment:
Is `fake_sydent` the recipient or the sender? Might be worth naming it as such?

Reply:
Fixed in https://github.com/matrix-org/sydent/pull/141/commits/d0db36137a09b28c86a24a6d22ed9bf037ae669e